File: 367-Valid-Perfect-Square.js
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function (num) {
    let l = 1, r = num

    while (l <= r) {
        let m = Math.floor((l + r) / 2)

        if (num === Math.pow(m, 2)) return true
        else if (num < Math.pow(m, 2)) r = m - 1
        else l = m + 1
    }

    return false
};

File: 367-Valid-Perfect-Square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        l, r = 0, num

        while l <= r:
            m = l + (r - l) // 2

            if m ** 2 == num:
                return True
            elif m ** 2 > num:
                r = m - 1
            else:
                l = m + 1
        
        return False